src/gpu-tips-n-tricks.md

```diff
@@ -1,61 +1,292 @@
+Instead of storing a pointer of type `T*`, each object that can be accessed on a device contains `Memory<T>` objects.
+The `Memory<T>` objects handle host and device pointers, memory allocations, and data synchronizations between host and device.
+
+To get the pointer `T*` from a `Memory<T>` object, one has to use the `Read()`, `Write()`, or `ReadWrite()` methods.
+
+-  `Read()` returns a `const T*` pointer, and should be used when the data will only be read,
+-  `Write()` returns a `T*` pointer, and should be used when writing data **without** using any previously contained data,
+-  `ReadWrite()` returns `T*` pointer, and should be used when read and write access to the pointer are required.
+
+`Read()`, `Write()`, and `ReadWrite()` are taking care automatically and completely transparently of data movements between host and device.
+
+The method `void UseDevice(bool)` specifies if a `Memory<T>` object is intended for computation on host or on device.
+The `Read()`,`Write()`, and `ReadWrite()` methods will return device pointer if using the device has been set to `true` with `UseDevice`, by default it is `false` and will return a host pointer.
+
+Sometimes, it is necessary to access data on host regardless, in this situation the `HostRead()`, `HostWrite()`, and `HostReadWrite()` methods should be used.
+
+In practice, developpers rarely have to manipulate `Memory<T>` objects, instead objects data can be stored using `Vector` and `Array<T>`.
+`Vector` and `Array<T>` data pointers can be accessed with the same methods as `Memory<T>`.
+```c++
+Vector v;
+v.UseDevice(true);
+const double *device_ptr = v.Read();
+const double *host_ptr = v.HostRead();
+```
+
+## The MFEM_FORALL macro
+The idea behind the `MFEM_FORALL` macro is to have the same behavior as a `for` loop and hide all the specifics to devices.
+Example:
+```c++
+for(int i = 0; i < N; i++)
+{
+  ...
+}
+```
+becomes
+```c++
+MFEM_FORALL(i, N,
+{
+  ...
+});
+```
+
+The `DeviceTensor<N,T>` class is an `N` dimension array containing elements of type `T`, `T` is `double` by default.
+It is convenient to use the `DeviceTensor<N,T>` class in combination with the memory manager and `MFEM_FORALL`.
+The `Reshape` function reshapes an array into an `N` dimension array:
+```c++
+Vector a;
+a.UseDevice(true);
+const int p = ...;
+const int q = ...;
+const int r = ...;
+const int N = ...;
+auto A = Reshape(a.Write(), p, q, r, N); // returns a DeviceTensor<4,double>
+MFEM_FORALL(n, N,
+{
+  for (int k = 0; k < r; k++)
+    for (int j = 0; j < q; j++)
+      for (int i = 0; i < p; i++)
+        A(i,j,k,n) = ...;
+});
+```
+There exists variants of this `MFEM_FORALL` macro, namely `MFEM_FORALL_2D` and `MFEM_FORALL_3D` which help maping 2D or 3D blocks of threads to the hardware more efficiently.
+In the case of a GPU, `MFEM_FORALL_3D(i,N,X,Y,Z,{...})` will declare `N` block of threads each of size `X`x`Y`x`Z` threads, whereas `MFEM_FORALL` uses `N` threads.
```

Comment:
I don't believe the wording here of saying `N` threads are being used is correct for this section. If we look at the MFEM_FORALL case, MFEM by default only launches 256 threads to run the CUDA kernels.

Reply:
I don't think this is correct, `MFEM_FORALL` launches N/256 blocks of 256 threads, so N threads. I think it would be confusing to describe this, the fact that we use 256 threads per block is an implementation detail in this case I think.

Reply:
Guess, it's been a while since I last looked at the actual documentation for how threading is done on CUDA devices, and I was wrong about my point earlier with the threads. However, you're still limited by the number of active concurrent threads that can be run at any one point: https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#differences-between-host-and-device . I know that number is hardware specific, so a max number shouldn't be put there. However, it might still be nice just to put a note in this section that all those threads may or may not be running concurrently depending on hardware limitations :)

Reply:
I see what you mean, but if I start talking about hardware threads, then I should probably talk about warps, SMs, etc... I agree that it is interesting and can be useful to understanding how to finely optimize code, but I think it would dive a bit too much into the hardware aspects of GPUs. I could add a note that by *thread* I always mean *software* thread, and never *hardware* thread.

Reply:
I think a note about what you mean by thread would work for me.

Reply:
Thank you very much for you feedback, do not hesitate to point everything you find unclear/confusing/insufficient.